core/sequencer/beam_fns.go

```diff
@@ -0,0 +1,223 @@
+	low := source.GetLowestWatermark()
+	high := source.GetHighestWatermark()
+	// Loop until less than batchSize items are returned.
+	for count := batchSize; count == batchSize; {
+		batch, err := s.logs.ReadLog(ctx, directoryID, logID, low, high, batchSize)
+		if err != nil {
+			return status.Errorf(codes.Internal, "ReadLog(): %v", err)
+		}
+		for _, m := range batch {
+			glog.Infof("emit(logID: %v, %v : EntryUpdate)", logID, m.ID)
+			emit(&ktpb.EntryUpdate{
+				Mutation:  m.Mutation,
+				Committed: m.ExtraData,
+			})
+			if m.ID > low {
+				low = m.ID
+			}
+		}
+		count = int32(len(batch))
+		glog.Infof("ReadLog(%v, (%v, %v], %v) count: %v", logID, low, high, batchSize, count)
+	}
+	return nil
+}
+
+// mapLogItem takes an individual entry and emits KV<index, mutation>
+// TODO(gbelvin): Make this mapping function part of the mutation interface.
+func mapLogItem(e *ktpb.EntryUpdate, emit func(index []byte, update *ktpb.EntryUpdate)) {
+	var entry ktpb.Entry
+	if err := proto.Unmarshal(e.GetMutation().GetEntry(), &entry); err != nil {
+		glog.Warningf("proto.Unmarshal failed: %v", err)
+	}
+	emit(entry.Index, e)
+}
+
+type mergeIndexFn struct{}
+
+func (*mergeIndexFn) CreateAccumulator() [][]byte                 { return [][]byte{} }
+func (*mergeIndexFn) AddInput(list [][]byte, val []byte) [][]byte { return append(list, val) }
+func (*mergeIndexFn) ExtractOutput(list [][]byte) [][]byte        { return list }
+func (*mergeIndexFn) MergeAccumulators(list [][][]byte) [][]byte {
+	ret := [][]byte{}
+	for _, l := range list {
+		ret = append(ret, l...)
+	}
+	return ret
+}
+
+// ReadMap queries the Trillian map for a list of leaves and emits KV<index, MapLeaf>
+func (s *Server) ReadMap(ctx context.Context, indexes [][]byte, directoryID string, rev int64,
+	emit func(index []byte, leaf *tpb.MapLeaf)) error {
+	// Fetch verification objects for directoryID.
+	config, err := s.ktServer.GetDirectory(ctx, &ktpb.GetDirectoryRequest{DirectoryId: directoryID})
+	if err != nil {
+		return err
+	}
+	mapClient, err := tclient.NewMapClientFromTree(s.tmap, config.Map)
+	if err != nil {
+		return err
+	}
+	leaves, err := mapClient.GetAndVerifyMapLeavesByRevision(ctx, rev, indexes)
+	if err != nil {
+		return err
+	}
+	for _, l := range leaves {
+		emit(l.Index, l)
+	}
+	return nil
+}
+
+// applyMutation processes all the mutations for a given index and emits the new map leaf.
+func applyMutation(index []byte,
+	getMapLeaf func(**tpb.MapLeaf) bool,
+	getMessage func(**ktpb.EntryUpdate) bool,
+	emit func(*tpb.MapLeaf)) error {
+	var oldMapLeaf *tpb.MapLeaf
+	if !getMapLeaf(&oldMapLeaf) {
+		return status.Errorf(codes.NotFound, "no map leaf found for index %x", index)
+	}
+	oldValue, err := entry.FromLeafValue(oldMapLeaf.LeafValue)
+	if err != nil {
+		glog.Warningf("entry.FromLeafValue(%v): %v", oldMapLeaf.LeafValue, err)
+		return err
+	}
+
+	var newEntry *ktpb.SignedEntry
+	var committed *ktpb.Committed
+	var msg *ktpb.EntryUpdate
+	for getMessage(&msg) {
```

Comment:
The order of entries returned from this function seems not deterministic to me, which makes the output of this pipeline not idempotent.

Reply:
Ah, so line 141 resolves the conflict, right? Sorry, seems like I just proved to myself that this code works, and could totally omit writing these comments :)

The only outstanding question is how okay is it to silently drop a mutation just because there is a conflicting one? Does this sort of thing need to be logged / monitored?